src/js/node/http2.ts

```diff
@@ -7,27 +7,963 @@ const { hideFromStack, throwNotImplemented } = require("internal/shared");
+  if (response !== undefined) response.emit("drain");
+}
+
+function onStreamAbortedRequest() {
+  const request = this[kRequest];
+  if (request !== undefined && request[kState].closed === false) {
+    request[kAborted] = true;
+    request.emit("aborted");
+  }
+}
+
+function resumeStream(stream) {
+  stream.resume();
+}
+
+function onStreamTrailersReady() {
+  this.sendTrailers(this[kResponse][kTrailers]);
+}
+
+function onStreamCloseResponse() {
+  const res = this[kResponse];
+
+  if (res === undefined) return;
+
+  const state = res[kState];
+
+  if (this.headRequest !== state.headRequest) return;
+
+  state.closed = true;
+
+  this.removeListener("wantTrailers", onStreamTrailersReady);
+  this[kResponse] = undefined;
+  res.emit("finish");
+
+  res.emit("close");
+}
+function onStreamCloseRequest() {
+  const req = this[kRequest];
+
+  if (req === undefined) return;
+
+  const state = req[kState];
+  state.closed = true;
+
+  req.push(null);
+  // If the user didn't interact with incoming data and didn't pipe it,
+  // dump it for compatibility with http1
+  if (!state.didRead && !req._readableState.resumeScheduled) req.resume();
+
+  this[kRequest] = undefined;
+
+  req.emit("close");
+}
+
+function onStreamTimeout() {
+  this.emit("timeout");
+}
+
+function isPseudoHeader(name) {
+  switch (name) {
+    case HTTP2_HEADER_STATUS: // :status
+    case HTTP2_HEADER_METHOD: // :method
+    case HTTP2_HEADER_PATH: // :path
+    case HTTP2_HEADER_AUTHORITY: // :authority
+    case HTTP2_HEADER_SCHEME: // :scheme
+      return true;
+    default:
+      return false;
+  }
+}
+
+function isConnectionHeaderAllowed(name, value) {
+  return name !== HTTP2_HEADER_CONNECTION || value === "trailers";
+}
+let statusConnectionHeaderWarned = false;
+let statusMessageWarned = false;
+function statusMessageWarn() {
+  if (statusMessageWarned === false) {
+    process.emitWarning("Status message is not supported by HTTP/2 (RFC7540 8.1.2.4)", "UnsupportedWarning");
+    statusMessageWarned = true;
+  }
+}
+
+function connectionHeaderMessageWarn() {
+  if (statusConnectionHeaderWarned === false) {
+    process.emitWarning(
+      "The provided connection header is not valid, " +
+        "the value will be dropped from the header and " +
+        "will never be in use.",
+      "UnsupportedWarning",
+    );
+    statusConnectionHeaderWarned = true;
+  }
+}
+
+function assertValidHeader(name, value) {
+  if (name === "" || typeof name !== "string" || StringPrototypeIncludes(name, " ")) {
+    throw new ERR_INVALID_HTTP_TOKEN("Header name", name);
+  }
+  if (isPseudoHeader(name)) {
+    throw new ERR_HTTP2_PSEUDOHEADER_NOT_ALLOWED();
+  }
+  if (value === undefined || value === null) {
+    throw new ERR_HTTP2_INVALID_HEADER_VALUE(value, name);
+  }
+  if (!isConnectionHeaderAllowed(name, value)) {
+    connectionHeaderMessageWarn();
+  }
+}
+
+hideFromStack(assertValidHeader);
+
+class Http2ServerRequest extends Readable {
```

Comment:
Is this a class in Node?

Reply:
It is https://github.com/nodejs/node/blob/4dc76a990f0037401b6bf8c982d7401c2533f23b/lib/internal/http2/compat.js#L311-L312